### PR TITLE
[Issue #2882] Setup logging in analytics codebase

### LINKED ISSUE #2882: Setup logging in analytics code
### Summary

Use the logic setup in the API code as a baseline, and adjust as necessary to fit with the analytics setup (ie. remove all the flask specific stuff).

### Acceptance criteria

- [ ] Logging configured on analytics code
- [ ] Tests working that validate logging

### PR DESCRIPTION
**Pull request by chouinar**
## Summary
Fixes #2882

## Changes proposed
Logging - a lot of it - based on what the API already does with some stuff trimmed out

Highlights:
* Python logging configured and usable with minimal effort
* Locally logging defaults to a human-readable format with some coloring, non-locally written out as JSON
* PII (SSN/TIN) auto-masking in all parts of a log
* Formatting of logs is ready for easy querying in Cloudwatch and New Relic
* Automatically append basic information to all log messages (eg. environment, task name)
* Automatically fetch basic ECS task information and attach to log messages

## Context for reviewers
This is ~90% copy-paste from the API code base with two notable exceptions:
* Linting rules are much stricter in analytics, so adjustments to formatting were made. Mostly docstring related
* The API has additional logic for things like request/response logging and other API and/or Flask specific concepts.

Assuming the new `init` method added to the cli file is always called (as we add more commands, probably want to split that into a few files), then logging will work anywhere. You simply need to do:

```
import logging

logger = logging.getLogger(__name__)
```
at the top of your file and that logger will _just work_. You can add extra values in the form of a dict that will get added to the overall JSON output for a log message.

Note that in order to get the ECS task stuff a decorator needs to be added to the entrypoint method of your script which I just added to the initialize_database command for the moment.

## Additional information
An example of the JSON logs generated by the task automatically (in this case, the last message saying it was successful):
```json
{
  "name": "analytics.logs.ecs_background_task",
  "msg": "Completed ECS task %s",
  "args": [
    "initialize_database"
  ],
  "levelname": "INFO",
  "levelno": 20,
  "pathname": "/analytics/src/analytics/logs/ecs_background_task.py",
  "filename": "ecs_background_task.py",
  "module": "ecs_background_task",
  "exc_info": null,
  "exc_text": null,
  "stack_info": null,
  "lineno": 82,
  "funcName": "_ecs_background_task_impl",
  "created": 1731706528.165175,
  "msecs": 165,
  "relativeCreated": 4169.229745864868,
  "thread": 281473801429024,
  "threadName": "MainThread",
  "processName": "MainProcess",
  "process": 1,
  "taskName": null,
  "ecs_task_duration_sec": 0.148,
  "status": "success",
  "environment": null,
  "task_name": "initialize_database",
  "task_uuid": "136089d8-b765-459d-9f59-d7f8cd9c8323",
  "message": "Completed ECS task initialize_database"
}
```

An example of logs when there is an exception:
```json
{
  "name": "analytics.logs.ecs_background_task",
  "msg": "ECS task failed",
  "args": [],
  "levelname": "ERROR",
  "levelno": 40,
  "pathname": "/analytics/src/analytics/logs/ecs_background_task.py",
  "filename": "ecs_background_task.py",
  "module": "ecs_background_task",
  "exc_info": [
    "<class 'Exception'>",
    "An example error",
    "<traceback object at 0xffff8ab4ffc0>"
  ],
  "exc_text": "Traceback (most recent call last):\n  File \"/analytics/src/analytics/logs/ecs_background_task.py\", line 72, in _ecs_background_task_impl\n    yield\n  File \"/analytics/src/analytics/logs/ecs_background_task.py\", line 52, in wrapper\n    return f(*args, **kwargs)\n           ^^^^^^^^^^^^^^^^^^\n  File \"/analytics/src/analytics/cli.py\", line 275, in initialize_database\n    raise Exception(\"An example error\")\nException: An example error",
  "stack_info": null,
  "lineno": 77,
  "funcName": "_ecs_background_task_impl",
  "created": 1731706654.6125245,
  "msecs": 612,
  "relativeCreated": 4152.841091156006,
  "thread": 281473382309920,
  "threadName": "MainThread",
  "processName": "MainProcess",
  "process": 1,
  "taskName": null,
  "status": "error",
  "environment": null,
  "task_name": "initialize_database",
  "task_uuid": "ec23ba72-0c3a-4479-a075-5884518ebfe9",
  "message": "ECS task failed"
}
```

Here are a few screenshots of running the task with human-readable logs enabled (the default locally). Including one where an exception is raised.

Screenshot 2024-11-15 at 4 20 21 PM

Screenshot 2024-11-15 at 4 20 32 PM

Screenshot 2024-11-15 at 4 38 51 PM

For context, here is what the output looked like before these changes:
```
initializing database
done
```

### CODE REVIEW

File: analytics/src/analytics/logs/pii.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    if key in ALLOW_NO_MASK:
+        return value
+    return _mask_pii(value)
+
+
+def _mask_pii(value: Any | None) -> Any | None:
+    if TIN_RE.search(str(value)):
+        return TIN_RE.sub("*********", str(value))
+    return value
```
Comment (coilysiren) [inline analytics/src/analytics/logs/pii.py:103]: Oh! I would expected this to come from a library

File: analytics/tests/logs/test_pii.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        ("123456789 test", "********* test"),
+        ("test 123456789", "test *********"),
+        ("test 123456789 test", "test ********* test"),
+        ("test=999000000.", "test=*********."),
+        ("test=999000000,", "test=*********,"),
+        (999000000.5, 999000000.5),
+        ({"a": "x", "b": "999000000"}, "{'a': 'x', 'b': '*********'}"),
+    ],
+)
+def test_mask_pii(value: Any | None, expected: Any | None):
```
Comment (coilysiren) [inline analytics/tests/logs/test_pii.py:34]: 😆 type annotations who

**Review by coilysiren — APPROVED** [review]
🚀 

(you probably want to update the docstring with click in it, though)

File: analytics/src/analytics/logs/pii.py
Comment (widal001) [inline analytics/src/analytics/logs/pii.py:1]: Glad that we have some PII filtering/masking. Thanks for adding this!

**Review by widal001 — APPROVED** [review]
Working beautifully!

Screenshot 2024-11-18 at 10 31 30 AM

I didn't have time for a super close review of the code, but I know that y'all have been using it on the API codebase for a minute.

**Review by coilysiren — DISMISSED** [review]

**Review by coilysiren — APPROVED** [review]

**Pull request merged**